### PR TITLE
🌱 Runtime SDK catalog

### PR DESCRIPTION
**Pull request by sbueringer**
should be ready for merge

some text for the PR description verifier

### CODE REVIEW

File: .github/workflows/golangci-lint.yml [hunk 1]
```diff
@@ -19,4 +19,3 @@ jobs:
         with:
           version: v1.44.0
           working-directory: ${{matrix.working-directory}}
```
Comment (sbueringer) [inline .github/workflows/golangci-lint.yml:21]: Didn't work so I revert it

File: internal/runtime/catalog/catalog.go [hunk 2]
```diff
@@ -19,285 +19,315 @@ package catalog
...
+// Catalog contains mappings between Go types and the corresponding
+// GroupVersionHooks, metadata and OpenAPIDefinitions.
 type Catalog struct {
-	// scheme allow access api-machinery utilities for handling types and types conversions
+	// scheme allows access to api-machinery utilities to implement conversions of
+	// request and response types.
 	scheme *runtime.Scheme
 
-	// gvhToType and typeToGVH allows tracking relations between hook go types and corresponding gvh.
+	// gvhToType allows mapping a GroupVersionHook to the Go type.
```
Comment (fabriziopandini) [inline internal/runtime/catalog/catalog.go:39]: nit 
```suggestion
	// gvhToType allows mapping a GroupVersionHook to the RuntimeHook function.
```
Same in other places

File: internal/runtime/catalog/catalog.go [hunk 3]
```diff
@@ -19,285 +19,315 @@ package catalog
...
 	gvhToType map[GroupVersionHook]reflect.Type
+
+	// gvhToType allows mapping a Go type to the GroupVersionHook.
 	typeToGVH map[reflect.Type]GroupVersionHook
 
-	// gvhDescriptors allows tracking relations gvh and corresponding descriptors.
-	gvhDescriptors map[GroupVersionHook]hookDescriptor
+	// gvhToHookDescriptor allows mapping a GroupVersionHook to the hook descriptor.
```
Comment (fabriziopandini) [inline internal/runtime/catalog/catalog.go:45]: nit
```suggestion
	// gvhToHookDescriptor allows mapping a GroupVersionHook to the corresponding hook descriptor.
```

File: internal/runtime/catalog/catalog.go [hunk 4]
```diff
@@ -19,285 +19,315 @@ package catalog
...
+	// catalogName is the name of this catalog. It is set based on the stack of the New caller.
+	// This is useful for error reporting to indicate the origin of the Catalog.
+	catalogName string
 }
 
-// OpenAPIDefinitionsGetter defines the signature of a func returning Open API definitions for
-// request and response types of a given group version.
-// NOTE: the OpenAPIDefinitionsGetter func are automatically generated by openapi-gen.
-type OpenAPIDefinitionsGetter func(ref common.ReferenceCallback) map[string]common.OpenAPIDefinition
+// Hook is a marker interface for a hook.
```
Comment (fabriziopandini) [inline internal/runtime/catalog/catalog.go:58]: nit
```suggestion
// Hook is a marker interface for a hook; Hooks should be defined as a func (*Request runtime.Object, *Response runtime.Object)
```

File: internal/runtime/catalog/catalog.go [hunk 5]
```diff
@@ -19,285 +19,315 @@ package catalog
...
-	t := reflect.TypeOf(hook)
-	if t.Kind() != reflect.Ptr {
-		panic("All service must be pointers to structs.")
+	// Validate that hookFunc is a func.
+	t := reflect.TypeOf(hookFunc)
+	if t.Kind() != reflect.Func {
+		panic("Hook must be a func")
+	}
+	if t.NumIn() != 2 {
+		panic("Hook must have two input parameter")
```
Comment (fabriziopandini) [inline internal/runtime/catalog/catalog.go:125]: nit
```suggestion
		panic("Hook must have two input parameter (*Request runtime.Object, *Response runtime.Object)")
```

File: internal/runtime/catalog/catalog.go [hunk 6]
```diff
@@ -19,285 +19,315 @@ package catalog
...
-		gvhToType:                 map[GroupVersionHook]reflect.Type{},
-		typeToGVH:                 map[reflect.Type]GroupVersionHook{},
-		gvhDescriptors:            map[GroupVersionHook]hookDescriptor{},
-		openAPIDefinitionsGetters: map[schema.GroupVersion]OpenAPIDefinitionsGetter{},
-	}
-	return s
+// hookDescriptor is an internal data structure which holds
+// all information about a Hook.
+type hookDescriptor struct {
+	hookMeta *HookMeta
```
Comment (fabriziopandini) [inline internal/runtime/catalog/catalog.go:64]: What about
```suggestion
	metadata *HookMeta
```

File: internal/runtime/catalog/test/catalog_test.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+)
+
+var c = catalog.New()
+
+func init() {
+	_ = v1alpha1.AddToCatalog(c)
+	_ = v1alpha2.AddToCatalog(c)
+}
+
+func TestCatalog(t *testing.T) {
```
Comment (fabriziopandini) [inline internal/runtime/catalog/test/catalog_test.go:41]: Not for the POC but we definitely need more tests (not only the happy path) + some comment explaining what we are testing

Comment (sbueringer) [reply internal/runtime/catalog/test/catalog_test.go:41]: Absolutely!

File: internal/runtime/catalog/catalog.go (hunk 4 above)
Comment (sbueringer) [reply internal/runtime/catalog/catalog.go:58]: Adde slightly differently the funcs have to use the concrete types not runtime.Object (although those types have to implement those interfaces)

File: internal/runtime/catalog/catalog_builder.go [hunk 8]
```diff
@@ -21,52 +21,56 @@ import (
...
-	// SchemeBuilder used to register functions for managing rpc method's request and
-	// response types.
-	// NOTE: this object is required to allow registration of automatically generated
-	// conversions func.
-	// TODO: investigate why func does not get registered if this is not a pointer
-	*runtime.SchemeBuilder
-}
+	// CatalogBuilder are the functions which are used
+	// to register Hooks and OpenAPIDefinitionsGetter into a catalog.
+	CatalogBuilder []func(*Catalog)
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/runtime/catalog/catalog_builder.go:33]: I exported it to stay consistent with the upstream scheme.Builder. 
But as we are currently not using it outside of the package I'll unexport this builder and the runtime.SchemeBuilder too.

We can always export later, vice versa is harder

**Pull request merged**